contracts/pangolin-roll/PangolinRoll.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.6.0;
+
+import "../pangolin-core/interfaces/IPangolinPair.sol";
+import "../pangolin-core/interfaces/IPangolinFactory.sol";
+import "../pangolin-periphery/interfaces/IERC20.sol";
+import "../pangolin-periphery/interfaces/IPangolinRouter.sol";
+import "../pangolin-periphery/libraries/PangolinLibrary.sol";
+import "./libraries/SafeERC20.sol";
+import "./libraries/Ownable.sol";
+
+// PangolinRoll helps your migrate your existing Sushiswap LP tokens to Pangolin LP ones
+contract PangolinRoll is Ownable {
+    using SafeERC20 for IERC20;
+
+    IPangolinRouter public oldRouter;
+    IPangolinRouter public router;
+    IERC20 public pngToken;
+    mapping(address => bool) public checkList;
+
+    constructor(
+        IPangolinRouter _oldRouter,
+        IPangolinRouter _router,
+        IERC20 _pngToken
+    ) public {
+        oldRouter = _oldRouter;
+        router = _router;
+        pngToken = _pngToken;
+    }
+
+    function migrateWithPermit(
+        address tokenA,
+        address tokenB,
+        uint256 liquidity,
+        uint256 amountAMin,
+        uint256 amountBMin,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public {
+        IPangolinPair pair = IPangolinPair(pairForOldRouter(tokenA, tokenB));
+        pair.permit(msg.sender, address(this), liquidity, deadline, v, r, s);
+
+        migrate(tokenA, tokenB, liquidity, amountAMin, amountBMin, deadline);
+    }
+
+    // msg.sender should have approved 'liquidity' amount of LP token of 'tokenA' and 'tokenB'
+    function migrate(
+        address tokenA,
+        address tokenB,
+        uint256 liquidity,
+        uint256 amountAMin,
+        uint256 amountBMin,
+        uint256 deadline
+    ) public {
+        require(deadline >= block.timestamp, "PangolinSwap: EXPIRED");
+
+        // Remove liquidity from the old router with permit
+        (uint256 amountA, uint256 amountB) = removeLiquidity(
+            tokenA,
+            tokenB,
+            liquidity,
+            amountAMin,
+            amountBMin
+        );
+
+        // Add liquidity to the new router
+        (uint256 pooledAmountA, uint256 pooledAmountB) = addLiquidity(
+            tokenA,
+            tokenB,
+            amountA,
+            amountB
+        );
+
+        // Send remaining tokens to msg.sender
+        if (amountA > pooledAmountA) {
+            IERC20(tokenA).safeTransfer(msg.sender, amountA - pooledAmountA);
+        }
+        if (amountB > pooledAmountB) {
+            IERC20(tokenB).safeTransfer(msg.sender, amountB - pooledAmountB);
+        }
+
+        // Transfer user a single PNG token if there are any remaining and user has not received one yet
+        if (address(pngToken) != address(0)) {
+            uint256 pngSupply = pngToken.balanceOf(address(this));
+            if (pngSupply > 0 && !checkList[msg.sender]) {
```

Comment:
This should check if the supply is greater than 1e18 or we risk bricking the contract if anybody sends PNG to it where the balance is between [0, 1]